### ISSUE TITLE
Better Native ICA support

### ISSUE THREAD

**Issue by JakeHartnell** [CosmWasm/wasmd#947]
Use case: _I want to write a contract that stakes an asset on a non-CosmWasm chain._

Really love implementing ICA in pure CosmWasm as with Confio's [cw-ibc-demo](https://github.com/confio/cw-ibc-demo). Unfortunately this will not work with non-CosmWasm chains like the Cosmos Hub.

Reading through the latest [ICA documentation](http://tutorials.cosmos.network/academy/4-ibc/ica.html), it seems we may need to call [RegisterInterchainAccount](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/host/keeper/account.go) and then can call `SendTx`? Or perhaps just `SendTx` is enough and it's smart enough to register the account?

Would be nice if we could extend `cosmwasm_std` [IbcMsg](https://docs.rs/cosmwasm-std/1.0.0/cosmwasm_std/enum.IbcMsg.html) to support native ICA messages. I'm hacking a bit today on getting them working using `CosmosMsg::Stargate`, but not sure how successful that will be and that it could be a much better dev experience to just support it outright.

In `cosmwasm_std` it would be ultimately nice to have something like:
```rust
#[non_exhaustive]
pub enum IbcMsg {
    ...
    IcaTx {
        channel_id: String,
        msgs: Vec<CosmosMsg<Empty>,
        timeout: IbcTimeout,
    }
    // Guess we need this to get the channel ID?
    RegisterIca {
        connnection_id: String,
        port_id: String,
        address: String
    }
}
```

Realize this issue also touches the `cosmwasm` repo, but figured it would be best to start the conversation here.

**Comment by ethanfrey** [comment]
This has been blocked on two items:

1. cosmos/interchain-accounts being officially promoted to the standard controller, so we can assume almost all chains use this. (it was currently marked demo and not to use in production)

2. the addition of callbacks to the go implementation. when you send a message to another chain, there is a good chance it failed. we need to inform the calling contract so it can handle that

I do believe IG is working on both of those items along with p2pvalidator. I have not followed their work the last month, but when it is ready, I am happy to look at this issue. (and add ics20 callbacks for success/failure in transfer)

**Comment by jtieri** [comment]
> Reading through the latest [ICA documentation](http://tutorials.cosmos.network/academy/4-ibc/ica.html), it seems we may need to call [RegisterInterchainAccount](https://github.com/cosmos/ibc-go/blob/main/modules/apps/27-interchain-accounts/host/keeper/account.go) and then can call `SendTx`? Or perhaps just `SendTx` is enough and it's smart enough to register the account?

`RegisterInterchainAccount` will need to be called first, this is where the logic lives for ensuring each native account on a controller chain maps 1:1 with an ICA on the host chain. This also initiates the channel handshake afaiu so it's necessary before you can send packets over the wire.

**Comment by giansalex** [comment]
I forked [interchain-accounts-demo](https://github.com/cosmos/interchain-accounts-demo/) and modified IBC middleware to call x/wasm because wasm does not support the ica-controller port.

For ica-controller only uses IBC methods: 
- `OnChanOpenInit`
- `OnChanOpenAck`
- `OnChanOpenAck`
- `OnAcknowledgementPacket`

IBC port format `icacontroller-{contract-address}`

The contract can call RegisterAccount, SubmitTx using CosmosMsg::Stargate and handle IBC Ack, Timeout.

> demo: Rcv and Ack txs
![image](https://user-images.githubusercontent.com/14926587/185728762-5b8c14f9-aea1-40e5-b207-eb8e50bd7279.png)

https://github.com/disperze/wasm-ica-demo/blob/wasm-ibc3/x/inter-tx/ibc_module.go
https://github.com/disperze/Juno/blob/ica-demo/app/app.go

could this be a way?

**Comment by faddat** [comment]
I have an unconfirmed hunch that this:

https://github.com/CosmWasm/wasmd/pull/938

and/ or this:

https://github.com/CosmWasm/wasmd/pull/940

get us somehow closer.  

@giansalex code is likely still useful, and maybe should be upstreamed into inter-tx directly.

**Comment by ethanfrey** [comment]
Put this on 0.32 milestone. Hopefully with cosmos-sdk 0.47 and ibc-go (v8?) we will have a standardised controller with proper callbacks as to whether ICA packets succeeded or failed. Until then no point to add support

**Comment by giansalex** [comment]
cw-ica-controller is live on JUNO mainnet 

`JUNO (controller) <> EVMOS (host)`
https://lcd.juno.disperze.network/ibc/core/channel/v1/channels/channel-207/ports/wasm.juno1xscn857dny0n9yxnf6macnnkgkfslahnue7jw84dgra4xf73efdq6nppnq

Code will be available with JUNO v12 upgrade, a small requirement depends on this upgrade.

**Comment by alpe** [comment]
ICA has been used with ibc-go for some time now. We need ADR-8 callbacks or something similar to cover the process. I have commented on https://github.com/CosmWasm/cosmwasm/issues/1652#issuecomment-1519588301

**Comment by alpe** [comment]
Moving this to a future release. We are blocked on ibc-go providing the adr-8 standard. I did some work in https://github.com/CosmWasm/wasmd/pull/1368 to help

**Comment by srdtrk** [comment]
It might be relevant to put this here, but it is currently possible to build a cosmwasm ica-controller that can communicate with the golang ica-host. This can handle callbacks too. I made such an implementation complete with end to end tests. Does it help resolve this issue? https://github.com/srdtrk/cw-ica-controller

**Comment by giansalex** [comment]
> This contract doesn't use any api to access ICA. It completes the handshake as the ica-controller. So your entire wasm chain could have ICA feature disabled but this implementation would still work.

Yes, following the philosophy of cw20-ics20.
Previously it was not possible due to a port name validation in ibc-go (counterparty chain), which was removed in v4.2

An implementation of ica-controller for liquid staking has been running smoothly.
![image](https://github.com/CosmWasm/wasmd/assets/14926587/9ba8b8b0-6691-4ea3-98e9-1e7900a495ea)

**Comment by srdtrk** [comment]
Good to know @giansalex. Is there a GitHub repo so that I can compare the implementations?
